lib/spack/spack/util/compression.py

```diff
@@ -59,21 +124,124 @@ def _unzip(archive_file):
+                ar.write(lar.read())
+    else:
+        if is_windows:
+            return _7zip(archive_file)
+        else:
+            return _xz(archive_file)
+
 
+def _xz(archive_file):
+    """Decompress lzma compressed .xz files via xz command line
+    tool. Available only on Unix
+    """
+    if is_windows:
+        raise RuntimeError('XZ tool unavailable on Windows')
+    _, ext = os.path.splitext(archive_file)
+    outfile = os.path.basename(archive_file.strip(ext))
+    xz = which('xz', required=True)
+    xz.add_default_arg('-d')
+    xz(archive_file)
+    return outfile
+
+
+def _7zip(archive_file):
+    """Unpack/decompress with 7z executable
+    7z is able to handle a number file extensions however
+    it may not be available on system.
+
+    Without 7z, Windows users with certain versions of Python may
+    be unable to extract .xz files, and all Windows users will be unable
+    to extract .Z files. If we cannot find 7z either externally or a
+    Spack installed copy, we fail, but inform the user that 7z can
+    be installed via `spack install 7zip`
+
+    Args:
+        archive_file (str): absolute path of file to be unarchived
+    """
+    _, ext = os.path.splitext(archive_file)
+    outfile = os.path.basename(archive_file.strip(ext))
+    tmp_src = os.path.join(os.getcwd(), 'tmp-src')
+    _7z = which('7z')
+    if not _7z:
+        raise CommandNotFoundError("7z unavailable,\
+unable to extract %s files. 7z can be installed via Spack" % ext)
+    _7z.add_default_arg('e')
+    _7z(archive_file)
+    return outfile
+
+
+def decompressor_for(path, ext=None):
+    """Wrapper for select_decompressor_for, returns
+    a function pointer to appropriate decompression
+    algorithm.
+
+    On Unix - simply invokes select_decompressor_for
 
-def decompressor_for(path, extension=None):
-    """Get the appropriate decompressor for a path."""
-    if ((extension and re.match(r'\.?zip$', extension)) or
+    On Windows - archives with one extension are
+    passed through to select_decompressor_for. Files
+    with two extensions (i.e. .tar.gz) are decomposed
+    into their components (.tar and .gz) and the results
+    of handling each extension are fed into the handler of
+    the next.
+
+    Args:
+        path (str): path of the archive file requiring decompression
+        ext (str): Extension of archive file
+    """
+    if ext:
+        if not allowed_archive(ext):
+            raise CommandNotFoundError("Unrecognized file extension")
+        if ((ext and re.match(r'\.?zip$', ext)) or
             path.endswith('.zip')):
-        return _unzip
-    if extension and re.match(r'gz', extension):
-        return _gunzip
-    if extension and re.match(r'bz2', extension):
-        bunzip2 = which('bunzip2', required=True)
-        return bunzip2
-    tar = which('tar', required=True)
-    tar.add_default_arg('-oxf')
-    return tar
+            return _unzip
+
+        if ext and re.match(r'gz', ext):
```

Comment:
Apologies can you add a note that `.tar.gz` files will not pass this check (since it's using `re.match` vs. `re.search`) - that was not clarified before this PR, but IMO it would be useful for future readers.

Reply:
Amended.